Fix mem corruption in `btree::base` dtor

ohamel-softwaresecure opened the pull request:
`btree::base` (or at least `btree::node`) needs a virtual dtor as it is `delete`d via pointer in `btree::clear`. 

This was picked up by asan whilst running the test suite. Perhaps ASAN/UBSAN should be enabled globally when running the test suite?

b-scholz commented:
The linter complained:
```
   */
         base(bool inner) : parent(nullptr), numElements(0), position(0), inner(inner) {}
 
-        virtual ~base() {}
+        virtual ~base() = default;
 
         bool isLeaf() const {
             return !inner;
@@ -379,7 +379,7 @@ protected:
         node(bool inner) : base(inner) {}
 
         // a destructor cleaning up nodes
-        ~node() {
+        ~node() override {
             if (this->inner) {
                 asInnerNode().cleanup();
             }
```

mmcgr commented:
This isn't a straightforward memory corruption issue. There isn't any memory corruption/loss from this, and the lack of a vtable improves performance.

Except, I don't find any specification that says this approach is okay - it's reliant on undocumented behaviour of new and delete. In my tests in the past there was no issue with memory, but this could change. It's also entirely possible for performance to be similar. I had started running performance comparisons for memory use and speed, but the results were questionable and I haven't had the time to retest in more detail.

The performance comparison should be done before this change is made. If corruption is actually shown, rather than the possibility warned of, then that would also be a great reason to change.

... and I typed too slow. We'll find out soon where performance is a problem.

b-scholz commented:
I feel it is a semantic issue if we have a dynamic class hierarchy and we don't have virtual destructors. We may need to resort to other styles CRTP etc to achieve similar behavior in a static fashion. 

Ok - let's check performance first whether we take a hit from this.

ohamel-softwaresecure commented:
Perhaps it would be worth using a custom allocator for btree so we can do this sort of trick legally?

ohamel-softwaresecure commented:
A possible hack/fix is to have user-defined new/delete operator that do the is-leaf check, casts `this` appropriately, and then delegates to malloc/free. I don't honestly know if this is completely legal, but it feels like it should be legal in C (assuming single class inheritance/composition).

For reference, here is the `test-suite.log`:

```
=======================================================
   souffle 1.7.1-1100-g7d70f5b6d: src/test-suite.log
=======================================================

# TOTAL: 30
# PASS:  26
# SKIP:  0
# XFAIL: 0
# FAIL:  4
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test/btree_set_test
=========================

=================================================================
==3384==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x617000000080 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   712 bytes;
  size of the deallocated type: 256 bytes.
    #0 0x7f996034df45 in operator delete(void*, unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45)
    #1 0x558a6b226705 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::clear() BTree.h:1921
    #2 0x558a6b218139 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::~btree() BTree.h:1282
    #3 0x558a6b21344d in souffle::btree_set<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, souffle::detail::comparator<int>, souffle::detail::updater<int> >::~btree_set() BTree.h:2214
    #4 0x558a6b209099 in souffle::test::test_BTreeSet_ParallelScaling::run() test/btree_set_test.cpp:822
    #5 0x558a6b1d5074 in main test/test.h:224
    #6 0x7f995fe0e1e2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x271e2)
    #7 0x558a6b1d4ced in _start (/home/olivier/projects/souffle/src/test/btree_set_test+0xcced)

0x617000000080 is located 0 bytes inside of 712-byte region [0x617000000080,0x617000000348)
allocated by thread T0 here:
    #0 0x7f996034c867 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x10f867)
    #1 0x558a6b2516d5 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node::grow_parent(souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node**, souffle::OptimisticReadWriteLock&, souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*, std::vector<souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*, std::allocator<souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*> >&) BTree.h:740
    #2 0x558a6b246d99 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node::split(souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node**, souffle::OptimisticReadWriteLock&, int, std::vector<souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*, std::allocator<souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*> >&) BTree.h:594
    #3 0x558a6b238e22 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node::rebalance_or_split(souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node**, souffle::OptimisticReadWriteLock&, int, std::vector<souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*, std::allocator<souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node*> >&) BTree.h:709
    #4 0x558a6b2212ef in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::insert(int const&, souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 256u, souffle::detail::linear_search, true, souffle::detail::comparator<int>, souffle::detail::updater<int> >::btree_operation_hints<1u>&) BTree.h:1525
    #5 0x558a6b210f6c in souffle::test::test_BTreeSet_ParallelScaling::run() [clone ._omp_fn.0] test/btree_set_test.cpp:835
    #6 0x7f9960003e95 in GOMP_parallel (/lib/x86_64-linux-gnu/libgomp.so.1+0x11e95)
    #7 0x7ffe58f0adaf  ([stack]+0x1cdaf)

SUMMARY: AddressSanitizer: new-delete-type-mismatch (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45) in operator delete(void*, unsigned long)
==3384==HINT: if you don't care about these errors you may set ASAN_OPTIONS=new_delete_type_mismatch=0
==3384==ABORTING
FAIL test/btree_set_test (exit status: 1)

FAIL: test/btree_multiset_test
==============================

BTreeMultiSet
	OK (4/4)	Clear
=================================================================
==3388==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x607000000100 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   72 bytes;
  size of the deallocated type: 40 bytes.
    #0 0x7f95dd776f45 in operator delete(void*, unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45)
    #1 0x559b32083dd3 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, false, souffle::detail::comparator<int>, souffle::detail::updater<int> >::clear() BTree.h:1921
    #2 0x559b320810f9 in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, false, souffle::detail::comparator<int>, souffle::detail::updater<int> >::~btree() BTree.h:1282
    #3 0x559b3207e76b in souffle::btree_multiset<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, souffle::detail::comparator<int>, souffle::detail::updater<int> >::~btree_multiset() BTree.h:2276
    #4 0x559b32067b58 in souffle::test::test_BTreeMultiSet_Load::run() test/btree_multiset_test.cpp:353
    #5 0x559b32050f54 in main test/test.h:224
    #6 0x7f95dd2711e2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x271e2)
    #7 0x559b32050bcd in _start (/home/olivier/projects/souffle/src/test/btree_multiset_test+0x9bcd)

0x607000000100 is located 0 bytes inside of 72-byte region [0x607000000100,0x607000000148)
allocated by thread T0 here:
    #0 0x7f95dd775867 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x10f867)
    #1 0x559b3209beac in souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, false, souffle::detail::comparator<int>, souffle::detail::updater<int> >::node* souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, false, souffle::detail::comparator<int>, souffle::detail::updater<int> >::buildSubTree<__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > >(__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > const&, __gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > const&) BTree.h:2164
    #2 0x559b3208d5b0 in std::enable_if<std::is_same<std::iterator_traits<__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > >::iterator_category, std::random_access_iterator_tag>::value, souffle::btree_multiset<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, souffle::detail::comparator<int>, souffle::detail::updater<int> > >::type souffle::detail::btree<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, false, souffle::detail::comparator<int>, souffle::detail::updater<int> >::load<souffle::btree_multiset<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, souffle::detail::comparator<int>, souffle::detail::updater<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > >(__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > const&, __gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > const&) BTree.h:2082
    #3 0x559b32083c5b in souffle::btree_multiset<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, souffle::detail::comparator<int>, souffle::detail::updater<int> > souffle::btree_multiset<int, souffle::detail::comparator<int>, std::allocator<int>, 16u, souffle::detail::linear_search, souffle::detail::comparator<int>, souffle::detail::updater<int> >::load<__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > >(__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > const&, __gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > > const&) BTree.h:2320
    #4 0x559b32066b7a in souffle::test::test_BTreeMultiSet_Load::run() test/btree_multiset_test.cpp:353
    #5 0x559b32050f54 in main test/test.h:224
    #6 0x7f95dd2711e2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x271e2)

SUMMARY: AddressSanitizer: new-delete-type-mismatch (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45) in operator delete(void*, unsigned long)
==3388==HINT: if you don't care about these errors you may set ASAN_OPTIONS=new_delete_type_mismatch=0
==3388==ABORTING
FAIL test/btree_multiset_test (exit status: 1)

FAIL: test/binary_relation_test
===============================

EqRelTest
number of threads: 8
=================================================================
==3387==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x61300004b7c0 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   376 bytes;
  size of the deallocated type: 256 bytes.
    #0 0x7ff74cc72f45 in operator delete(void*, unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45)
    #1 0x5619bbdbb825 in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::inner_node::cleanup() BTree.h:1072
    #2 0x5619bbdb419b in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node::~node() BTree.h:382
    #3 0x5619bbdbb818 in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::inner_node::cleanup() BTree.h:1072
    #4 0x5619bbdb419b in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node::~node() BTree.h:382
    #5 0x5619bbdbb818 in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::inner_node::cleanup() BTree.h:1072
    #6 0x5619bbdb419b in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node::~node() BTree.h:382
    #7 0x5619bbdac77c in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::clear() BTree.h:1921
    #8 0x5619bbda4c99 in souffle::EquivalenceRelation<souffle::Tuple<int, 2ul> >::emptyPartition() const EquivalenceRelation.h:196
    #9 0x5619bbd9ddd5 in souffle::EquivalenceRelation<souffle::Tuple<int, 2ul> >::~EquivalenceRelation() EquivalenceRelation.h:51
    #10 0x5619bbd9425f in souffle::test::test_EqRelTest_ParallelScaling::run() test/binary_relation_test.cpp:552
    #11 0x5619bbd73073 in main test/test.h:224
    #12 0x7ff74c7331e2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x271e2)
    #13 0x5619bbd72ccd in _start (/home/olivier/projects/souffle/src/test/binary_relation_test+0x9ccd)

0x61300004b7c0 is located 0 bytes inside of 376-byte region [0x61300004b7c0,0x61300004b938)
allocated by thread T0 here:
    #0 0x7ff74cc71867 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x10f867)
    #1 0x5619bbdc6cf5 in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node::grow_parent(souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node**, souffle::OptimisticReadWriteLock&, souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*, std::vector<souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*, std::allocator<souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*> >&) BTree.h:740
    #2 0x5619bbdc23eb in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node::split(souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node**, souffle::OptimisticReadWriteLock&, int, std::vector<souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*, std::allocator<souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*> >&) BTree.h:594
    #3 0x5619bbdbcf3a in souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node::rebalance_or_split(souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node**, souffle::OptimisticReadWriteLock&, int, std::vector<souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*, std::allocator<souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::node*> >&) BTree.h:709
    #4 0x5619bbdb603c in souffle::detail::LambdaBTree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, std::function<souffle::PiggyList<int>* (std::pair<int, souffle::PiggyList<int>*>&)>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::insert(std::pair<int, souffle::PiggyList<int>*>&, souffle::detail::btree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::btree_operation_hints<1u>&, std::function<souffle::PiggyList<int>* (std::pair<int, souffle::PiggyList<int>*>&)> const&) LambdaBTree.h:318
    #5 0x5619bbdacd19 in souffle::detail::LambdaBTree<std::pair<int, souffle::PiggyList<int>*>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, std::allocator<std::pair<int, souffle::PiggyList<int>*> >, 256u, souffle::detail::binary_search, true, std::function<souffle::PiggyList<int>* (std::pair<int, souffle::PiggyList<int>*>&)>, souffle::EqrelMapComparator<std::pair<int, souffle::PiggyList<int>*> >, souffle::detail::updater<std::pair<int, souffle::PiggyList<int>*> > >::insert(std::pair<int, souffle::PiggyList<int>*>&, std::function<souffle::PiggyList<int>* (std::pair<int, souffle::PiggyList<int>*>&)> const&) LambdaBTree.h:64
    #6 0x5619bbda5103 in souffle::EquivalenceRelation<souffle::Tuple<int, 2ul> >::genAllDisjointSetLists() const EquivalenceRelation.h:711
    #7 0x5619bbd9deaa in souffle::EquivalenceRelation<souffle::Tuple<int, 2ul> >::size() const EquivalenceRelation.h:216
    #8 0x5619bbd93e7f in souffle::test::test_EqRelTest_ParallelScaling::run() test/binary_relation_test.cpp:560
    #9 0x5619bbd73073 in main test/test.h:224
    #10 0x7ff74c7331e2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x271e2)

SUMMARY: AddressSanitizer: new-delete-type-mismatch (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45) in operator delete(void*, unsigned long)
==3387==HINT: if you don't care about these errors you may set ASAN_OPTIONS=new_delete_type_mismatch=0
==3387==ABORTING
FAIL test/binary_relation_test (exit status: 1)

FAIL: test/eqrel_datastructure_test
===================================

DjTest
	OK (10002/10002)	ParallelScaling
	OK (1/1)	Clear
	OK (10/10)	TestUnion
	OK (7/7)	MakeNode
	OK (0/0)	Scoping
LambdaBTree
number of times the functor has been called: 100
=================================================================
==3386==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x613000000040 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   376 bytes;
  size of the deallocated type: 256 bytes.
    #0 0x7f57873d1f45 in operator delete(void*, unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45)
    #1 0x55de7ce09dcf in souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::clear() BTree.h:1921
    #2 0x55de7ce05a5f in souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::~btree() BTree.h:1282
    #3 0x55de7ce00333 in souffle::detail::LambdaBTree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::~LambdaBTree() LambdaBTree.h:50
    #4 0x55de7ce00353 in souffle::LambdaBTreeSet<std::pair<unsigned long, unsigned long>, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search>::~LambdaBTreeSet() LambdaBTree.h:577
    #5 0x55de7cdf2301 in souffle::test::test_LambdaBTree_ContendParallel::run() test/eqrel_datastructure_test.cpp:646
    #6 0x55de7cdcf074 in main test/test.h:224
    #7 0x7f5786e921e2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x271e2)
    #8 0x55de7cdceced in _start (/home/olivier/projects/souffle/src/test/eqrel_datastructure_test+0xbced)

0x613000000040 is located 0 bytes inside of 376-byte region [0x613000000040,0x6130000001b8)
allocated by thread T3 here:
    #0 0x7f57873d0867 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.5+0x10f867)
    #1 0x55de7ce20053 in souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node::grow_parent(souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node**, souffle::OptimisticReadWriteLock&, souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*, std::vector<souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*, std::allocator<souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*> >&) BTree.h:740
    #2 0x55de7ce1b98b in souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node::split(souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node**, souffle::OptimisticReadWriteLock&, int, std::vector<souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*, std::allocator<souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*> >&) BTree.h:594
    #3 0x55de7ce1549e in souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node::rebalance_or_split(souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node**, souffle::OptimisticReadWriteLock&, int, std::vector<souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*, std::allocator<souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::node*> >&) BTree.h:709
    #4 0x55de7ce0df83 in souffle::detail::LambdaBTree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::insert(std::pair<unsigned long, unsigned long>&, souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::btree_operation_hints<1u>&, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)> const&) LambdaBTree.h:318
    #5 0x55de7ce0db9c in souffle::detail::LambdaBTree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::insert(std::pair<unsigned long, unsigned long>&, souffle::detail::btree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::btree_operation_hints<1u>&, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)> const&) LambdaBTree.h:276
    #6 0x55de7ce075fb in souffle::detail::LambdaBTree<std::pair<unsigned long, unsigned long>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<unsigned long, unsigned long> >, 256u, souffle::detail::binary_search, true, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)>, souffle::EqrelMapComparator<std::pair<unsigned long, unsigned long> >, souffle::detail::updater<std::pair<unsigned long, unsigned long> > >::insert(std::pair<unsigned long, unsigned long>&, std::function<unsigned long (std::pair<unsigned long, unsigned long>&)> const&) LambdaBTree.h:64
    #7 0x55de7cdfa4dc in souffle::test::test_LambdaBTree_ContendParallel::run() [clone ._omp_fn.0] test/eqrel_datastructure_test.cpp:670
    #8 0x7f578709131d  (/lib/x86_64-linux-gnu/libgomp.so.1+0x1b31d)

Thread T3 created by T0 here:
    #0 0x7f57872fb805 in pthread_create (/lib/x86_64-linux-gnu/libasan.so.5+0x3a805)
    #1 0x7f5787091985  (/lib/x86_64-linux-gnu/libgomp.so.1+0x1b985)
    #2 0x7f5787087e90 in GOMP_parallel (/lib/x86_64-linux-gnu/libgomp.so.1+0x11e90)
    #3 0x7f5787cf3a1f  (<unknown module>)
    #4 0xffff  (<unknown module>)

SUMMARY: AddressSanitizer: new-delete-type-mismatch (/lib/x86_64-linux-gnu/libasan.so.5+0x110f45) in operator delete(void*, unsigned long)
==3386==HINT: if you don't care about these errors you may set ASAN_OPTIONS=new_delete_type_mismatch=0
==3386==ABORTING
FAIL test/eqrel_datastructure_test (exit status: 1)

```

mmcgr commented:
Our standard benchmarking does show an increase in memory, and a more mixed increase in execution time, but the change is minor: http://plang1.it.usyd.edu.au/jenkins/job/Master-Performance/plot/#plot-8671179143402261283.csv . If this is a problem, then it will be with more extreme cases where memory is already limiting.

b-scholz commented:
I had a quick discussion with Martin. The best way forward would be to implement memory pools for inner nodes and leaves for the btree data-structure. Note that we don't have deletions and hence we can allocate large memory chunks which we can free in one go when the data-structure is deleted. Freeing large btrees is time-consuming which we discovered in previous experiments.

b-scholz commented:
I sketched a memory-pool manager in my branch:
https://github.com/b-scholz/souffle/tree/memorypool

It works only for sequential execution and more work is required:
(1) make it race free 
(2) use thread-local memory for memory chunks to minimize synchronization
(3) fix the following bug in growth_parent 

It crashes for this problem:
```
.decl A,B(x:number, y:number)

A(1,2).
A(1000, 1).
A(x+1,x+2) :- A(x,_),  x < 1000.

B(x,y) :- A(x,y).
B(x,z) :- B(x,y), A(y,z).

.printsize B
```
